modnet/sklearn.py

```diff
@@ -0,0 +1,134 @@
+  (see https://scikit-learn.org/stable/modules/compose.html#caching-transformers-avoid-repeated-computation).
+* Another note about the fit method of a Pipeline object:
+  It is possible to pass fit parameters to each step of a Pipeline object
+  (see https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html#sklearn.pipeline.Pipeline.fit)
+* Another note about the __init__ method of sklearn derived estimators:
+  The method should ONLY set the instance variables. In particular, "every keyword argument accepted by __init__
+  should correspond to an attribute on the instance" (as stated in sklearn's developer documentation)
+  (see https://scikit-learn.org/stable/developers/develop.html#instantiation).
+"""
+
+
+from sklearn.base import BaseEstimator
+from sklearn.base import RegressorMixin
+from sklearn.base import TransformerMixin
+from typing import Union
+from typing import Dict
+
+
+class MODNetFeaturizer(TransformerMixin, BaseEstimator):
+    """"""
+
+    def __init__(self, matminer_featurizers):
+        """Constructor for MODNetFeaturizer"""
+        self.matminer_featurizers = matminer_featurizers
+
+    def fit(self, X, y=None):
+        """Probably not needed except for some matminer featurizers (e.g. I think SOAP needs to be fitted before)."""
+        return self
+
+    def transform(self, X, y=None):
+        """Transform the input data (i.e. containing the composition and/or structure to the features"""
+
+    @classmethod
+    def from_preset(cls, preset):
+        """Initializes a MODNetFeaturizer class based on some preset.
+
+        Args:
+            preset: Name of the preset (e.g. "DeBreuck2020" could trigger the Structure+Composition featurizers)
+
+        Notes:
+            See in matminer how it is done, e.g. SiteStatsFingerprint.from_preset method.
+        """
+
+
+class RR(TransformerMixin, BaseEstimator):
+    """Relevance-Redundancy (RR) analysis.
+
+    Feature selection and ordering procedure based on the relevance-redundancy algorithm developed in ARTICLE.
+
+    Blabla
+
+    Notes:
+        1. Uses preprocessing.get_features_relevance_redundancy under the hood.
+        2. To check if it is ok to have y that is *NOT* ignored (i.e. not y=None) in the fit method. Usually a
+           Transformer's fit method signature is fit(self, X, y=None, ...).
+    """
+
+    def __init__(self, n_feat: Union[None, int]=None, rr_parameters: Union[None, Dict]=None):
+        """Constructor for RR transformer.
+
+        Args:
+            n_feat: Number of features to keep and reorder using the RR procedure (default: None, i.e. all features).
+            rr_parameters: Allow to tune p and c parameters. (default: None, i.e. use the dynamical setting in ARTICLE).
+
+        Notes:
+            This method should not be changed
+        """
+        self.n_feat = n_feat
+        self.rr_parameters = rr_parameters
+
+    def fit(self, X, y, nmi_feats_target=None, cross_nmi_feats=None):
```

Comment:
Best practice is that the user computes the nmi_target and nmi_cross separately, and stores it somewhere. The current API somewhat suggests the opposite by including this as part of the RR class, doing it on-the-fly.

Reply:
I would not say that the API suggests that it is best practice. It only allows to do it on the fly. The documentation and in particular the tutorial(s) should mention the best practice of precomputing nmi_target and nmi_cross and then using them in the fit method.

Reply:
I'm trying to work this into my PR on MODData at the moment, so that basically everything can be cached and reloaded (at the moment the API just allows you to pickle the entire MODData object) My changes are getting a bit out of hand locally, so I'll try to split it up into a couple of different PRs

Reply:
Great Mathew :)